SqlClient: Avoid intermediate TaskCompletionSources

justinvp opened the pull request:
Minor cleanup:

- Replace intermediate `TaskCompletionSource`s with calls to `Task.FromResult`, `Task.FromCanceled`, and `Task.FromException`.

- Replace `ADP.CreatedTaskWithException` and `ADP.CreatedTaskWithCancellation` with direct calls to `Task.FromException` and `Task.FromCanceled`.

cc: @saurabh500, @stephentoub

stephentoub commented:
The code changes LGTM.

I seem to remember, though, that we were aiming to have this assembly work pre-.NET 4.6, and I thought we'd made changes to enable that, but as this is successfully compiling while still using APIs introduced in .NET 4.6, I may be remembering incorrectly. @saurabh500? @ericstj?

justinvp commented:
> I seem to remember, though, that we were aiming to have this assembly work pre-.NET 4.6

Interesting. This assembly was already using `Task.FromException` in the implementation of `ADP.CreatedTaskWithException`, which was introduced by @saurabh500 (perhaps mistakenly?) while addressing feedback when adding these sources to corefx.

stephentoub commented:
Ok, thanks, I'm likely remembering incorrectly then.

stephentoub commented:
I think I was thinking of https://github.com/dotnet/corefx/issues/4802. LGTM.

ericstj commented:
If you look at the configurations of this project you see it cross compiles for desktop.  For older versions of desktop or is a facade and doesn't use the API you are pointing out.

stephentoub commented:
@ericstj, it appears that all of the files Justin is modifying are similarly in the ```<ItemGroup Condition="'$(IsPartialFacadeAssembly)' != 'true'">``` section, so this is fine, right?

saurabh500 commented:
@stephentoub this should be fine. The files changed are used when not compiling for .Net desktop.
In case of desktop the assembly targets System.Data and forwards the API calls to desktop assembly.

saurabh500 commented:
LGTM
